### PR TITLE
Use a single `installer-gather.sh` for bip, non-bip, pre-pivot and post-pivot with Control Plane gather

### PR DESCRIPTION
**Pull request by omertuc**
As discussed in the bootstrap-in-place PR comments

### CODE REVIEW

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (eranco74) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:136]: kubeconfig?

Comment (omertuc) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:136]: see `queue` function diff

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (eranco74) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:18]: Please add a comment here and in the else about when you expect this file to exist

Comment (omertuc) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:18]: b1c9eb40a

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (jstuever) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:16]: I'd prefer to see this generalized so the same process can be used by other workflows. Something like `log-bundle-bootstrap` would do.

Comment (jstuever) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:16]: Same for the variable.

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (jstuever) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:18]: I'm not a fan of this workflow. I would rather see some the typical steps fail when ran on a master node, and then a section where some files are cleaned up before extracting the tarball. ie
```
exec &> >(tee "${ARTIFACTS}/gather.log")
echo "Gathering bootstrap systemd summary ..."
echo "Gathering bootstrap containers ..."
echo "Gathering rendered assets..."
echo "Gathering bootstrap journals ..."
if [[ -f "/etc/kubernetes/bootstrap-secrets/kubeconfig"]]; then
    GATHER_KUBECONFIG=/etc/kubernetes/bootstrap-secrets/kubeconfig
fi
echo "Gathering cluster resources ..."
echo "Gathering bootstrap journals ..."
echo "Waiting for logs ..."
if [[ -f ${BOOTSTRAP_IN_PLACE_BOOTSTRAP_PHASE_ARCHIVE_PATH} ]]; then
   rm ${ARTIFACTS}/things
   tar -xzf ${BOOTSTRAP_IN_PLACE_BOOTSTRAP_PHASE_ARCHIVE_PATH}
fi
echo "Gather remote logs"
```

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (jstuever) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:21]: Keep this in gather.log

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (jstuever) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:26]: Some things from the bootstrap log bundle can probably be placed into ${ARTIFACTS}/bootstrap while others might need a new home such as ${ARTIFACTS}/bootstrap-log-bundle. For example, the "resources" from the bootstrap-log-bundle do not represent the current cluster state and might go better in /bootstrap-log-bundle subdir. And, the bootstrap/containers is already captured in control-plane/master/containers and can be deleted and replaced by those in the bootstrap-log-bundle.

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (jstuever) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:103]: This needs to stay at the bottom as `resources/masters.list` is populated by `Gathering cluster resources ...`. I'll leave a comment about reworking this logic below.

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (jstuever) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh:169]: "Gather remote logs" needs to be here. We can rework the logic
```
export MASTERS=()
if [[ -f ${BOOTSTRAP_IN_PLACE_BOOTSTRAP_PHASE_ARCHIVE_PATH} ]]; then
    # Instead of running installer-masters-gather.sh on remote masters, run it on ourselves
    MASTER_GATHER_ID="master-${GATHER_ID}"
    MASTER_ARTIFACTS="/tmp/artifacts-${MASTER_GATHER_ID}"
    mkdir -p "${ARTIFACTS}/control-plane/master"
    sudo /usr/local/bin/installer-masters-gather.sh --id "${MASTER_GATHER_ID}" </dev/null
    cp -r "$MASTER_ARTIFACTS"/* "${ARTIFACTS}/control-plane/master/"
elif [ "$#" -ne 0 ]; then
    MASTERS=( "$@" )
elif test -s "${ARTIFACTS}/resources/masters.list"; then
    mapfile -t MASTERS < "${ARTIFACTS}/resources/masters.list"
else
    echo "No masters found!"
fi
```

The rest should work 'as-is' as `MASTERS=()`

**Review by jstuever — CHANGES_REQUESTED** [review]
I believe this is getting us closer to a feasible solution. I have a few changes I'd like to see here in order to 1) reduce the confusion/complexity of the installer-gather.sh script itself, 2) ensure that the various sections of the script work as expected, and 3) reduce the differences in the resulting tarball(s).

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (omertuc) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:21]: You mean `>>` ?

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh
Comment (eranco74) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:26]: I like this direction and I think we can take it one step further https://github.com/eranco74/installer/pull/13.
I think that all things from the bootstrap log bundle should get a new home - ${ARTIFACTS}/${LOG_BUNDLE_NAME}.


I think this is the way to go for few reasons:

1. The log bundle collected while running from liveCD are defacto logs from a different node.
2. We won't lose any information (such as the crio service logs pre/post reboot).
3. It will simplify the code since there is no need for all the `rm` and `cp`, installer-gather.sh can stay the same and just add the log-bundle in case it exists.
4. This feature can work for any existing log bundle and not just for single node bootstrap-in-place.

While we want to have a similar log bundle to the regular installation flow we can't ignore that the bootstrap logic ran on "another node" earlier in the process, so instead of "hiding it" we should explicitly reflect that by placing it under a new home.

Comment (jstuever) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:26]: I agree, https://github.com/eranco74/installer/pull/13 looks very promising. Can we get that pulled into https://github.com/openshift/installer/pull/4482? I have a few more thoughts that I'll add there.

Comment (omertuc) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:26]: It'll be like 10 hours until @eranco74 will merge it so if you have comments specific to #13 please write them in #13 now so we can deal with them in our morning in 10h, because our timezones are out of sync

Comment (jstuever) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh:26]: Not specific to #13... but #13 might need some modificaitons. https://github.com/openshift/installer/pull/4482/files#r568957299

**Comment by omertuc** [comment]
obsolete

**Pull request closed without merging**
